main.cpp

```diff
@@ -108,7 +108,410 @@ write 3 UDTs below that EACH have:
+        ~Student();
+        void addGrade(double grade);
+        void printDetails() const;
+        double calculateAverageGrade() const;
+        void setFinalGrade();
+    };
+
+    std::vector<Student> students;
+    std::string classroomName;
+    int totalStudents = 0;
+    double averageGrade = 0;
+    char averageLetterGrade;
+
+    Classroom(std::string name);
+    ~Classroom();
+    void registerStudent(const Student& student);
+    static char convertGradeIntoLetterGrade(double grade);
+    void calculateAverageGrade();
+    void printClassDetails();
+};
+//--
+int Classroom::Student::studentId = 0;
+//--
+Classroom::Student::Student(std::string n, int a) : name(n), age(a)
+{
+    studentId++;
+}
+//--
+Classroom::Student::~Student()
+{
+    std::cout << "Student destructor called!" << std::endl;
+}
+//--
+void Classroom::Student::addGrade(double grade) 
+{
+    grades.push_back(grade);
+}
+//--
+void Classroom::Student::printDetails() const
+{
+    std::cout << "Student ID: " << studentId << ", Name: " << name << ", Age: " << age << std::endl;
+    std::cout << "Grades: ";
+    for (double grade : grades) 
+    {
+        std::cout << grade << " ";
+    }
+    std::cout << "\nAverage Grade: " << calculateAverageGrade() << std::endl;
+}
+//--
+void Classroom::Student::setFinalGrade()
+{
+    finalGrade = convertGradeIntoLetterGrade(Student::calculateAverageGrade());
+}
+//--
+double Classroom::Student::calculateAverageGrade() const{
+    if (grades.empty())
+    {
+        return 0.0;
+    }
+    double sum = 0.0;
+    for (double grade : grades) 
+    {
+        sum += grade;
+    }
+    return sum / static_cast<double>(grades.size());
+}
+//--
+Classroom::Classroom(std::string name) : 
+    classroomName(name)
+{
+}
+//--
+Classroom::~Classroom()
+{
+    std::cout << "Classroom destructor called!" << std::endl;
+}
+//--
+void Classroom::registerStudent(const Student& student) 
+{
+    students.push_back(student);
+    totalStudents++;
+}
+//--
+void Classroom::calculateAverageGrade()
+{
+    if (totalStudents == 0)
+    {
+        averageGrade = 0;
+        averageLetterGrade = 'F';
+        return;
+    }
+    double sum = 0.0;
+    for (const auto& student : students)
+    {
+        sum += student.calculateAverageGrade();
+    }
+    averageGrade = sum / totalStudents;
+    averageLetterGrade = convertGradeIntoLetterGrade(averageGrade);
+}
+//--
+void Classroom::printClassDetails(){
+    std::cout << "Classroom: " << classroomName << ", Total Students: " << totalStudents << std::endl;
+    std::cout << "Class Average Grade: " << averageGrade << " (" << averageLetterGrade << ")" << std::endl;
+    for (const Student& student : students) 
+    {
+        student.printDetails();
+    }
+}
+//--
+char Classroom::convertGradeIntoLetterGrade(double grade)
+{
+    if (grade >= 90) return 'A';
+    else if (grade >= 80) return 'B';
+    else if (grade >= 70) return 'C';
+    else if (grade >= 60) return 'D';
+    else return 'F';
+}
```

Comment:
If you find yourself writing code like this:
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    else 
    {
        return anotherValue;
    }
}
```
You do not need the `else` block.  Here's why:
`return` statements are the last thing that **ever happens** in a function.
When the `if()` expression is evaluated as `true` and the block is entered, the `return aValue;` is the last line of code to be executed before the function exits.
but when the `if()` expression evaluates as `false` and the block is **NOT** entered, the next thing that happens is `return anotherValue;`
you have no code after the `else` block that needs executing, since your `else` block contains a `return` statement.
that means you do **NOT** need the `else` block, and it is just causing visual clutter.
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    
    return anotherValue;
}
```
- [ ] revise